src/main/java/org/elasticsearch/plugins/repository/PluginDescriptor.java

```diff
@@ -0,0 +1,357 @@
+        }
+
+        public Builder jvm() {
+            this.jvm = true;
+            return this;
+        }
+
+        public Builder dependency(String fqn) {
+            dependencies.add(pluginDescriptor(fqn).build());
+            return this;
+        }
+
+        public Builder dependency(PluginDescriptor dependency) {
+            dependencies.add(dependency);
+            return this;
+        }
+
+        public Builder dependency(String organisation, String name, String version) {
+            PluginDescriptor dependency = pluginDescriptor(name).organisation(organisation).version(version).build();
+            dependencies.add(dependency);
+            return this;
+        }
+
+        public Builder loadFromProperties(InputStream is) throws IOException {
+            Properties props = new Properties();
+            props.load(is);
+
+            String version = props.getProperty("version");
+            if (Strings.hasText(version)) {
+                version(version);
+            }
+
+            String desc = props.getProperty("description");
+            if (Strings.hasText(desc)) {
+                description(desc);
+            }
+            return this;
+        }
+
+        public Builder loadFromProperties(Path esPlugin) {
+            try (InputStream is = Files.newInputStream(esPlugin)) {
+                return loadFromProperties(is);
+            } catch (IOException e) {
+                throw new ElasticsearchException("Error when loading plugin descriptor from properties file [" + esPlugin.toAbsolutePath() + "]", e);
+            }
+        }
+
+        public Builder loadFromMeta(Settings settings) {
+            String org = settings.get("organisation");
+            if (Strings.hasText(org)) {
+                organisation(org);
+            }
+            String name = settings.get("name");
+            if (Strings.hasText(name)) {
+                name(name);
+            }
+            String version = settings.get("version");
+            if (Strings.hasText(version)) {
+                version(version);
+            }
+            String description = settings.get("description");
+            if (Strings.hasText(description)) {
+                description(description);
+            }
+
+            ImmutableMap<String, String> deps = settings.getByPrefix("dependencies.").getAsMap();
+            if ((deps != null) && (!deps.isEmpty())) {
+                int i = 0;
+                String depName;
+
+                do {
+                    depName = deps.get(String.valueOf(i) + ".name");
+                    String depOrganisation = deps.get(String.valueOf(i) + ".organisation");
+                    String depVersion = deps.get(String.valueOf(i) + ".version");
+                    if (Strings.hasText(depName)) {
+                        dependency(depOrganisation, depName, depVersion);
+                    }
+                    i++;
+                } while ((depName != null) && (i < 1000));
```

Comment:
I'm not a big fan of `do {} while ();`. Don't know what others think though.
If you keep this, please write:

```java
} while (depName != null && i < 1000);
```